internal/datasystem/fdv2_datasystem.go

```diff
@@ -0,0 +1,253 @@
+	"sync"
+)
+
+var _ subsystems.DataSourceUpdateSink = (*store)(nil)
+
+type broadcasters struct {
+	dataSourceStatus *internal.Broadcaster[interfaces.DataSourceStatus]
+	dataStoreStatus  *internal.Broadcaster[interfaces.DataStoreStatus]
+	flagChangeEvent  *internal.Broadcaster[interfaces.FlagChangeEvent]
+}
+
+type FDv2 struct {
+	// Operates the in-memory and optional persistent store that backs data queries.
+	store *store
+
+	// List of initializers that are capable of obtaining an initial payload of data.
+	initializers []subsystems.DataInitializer
+
+	// The primary synchronizer responsible for keeping data up-to-date.
+	primarySync subsystems.DataSynchronizer
+
+	// The secondary synchronizer, in case the primary is unavailable.
+	secondarySync subsystems.DataSynchronizer
+
+	// Whether the SDK should make use of persistent store/initializers/synchronizers or not.
+	offline bool
+
+	loggers ldlog.Loggers
+
+	// Cancel and wg are used to track and stop the goroutines used by the system.
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+
+	// The SDK client, via MakeClient, expects to pass a channel down into a data source which will then be
+	// closed when the source is considered to be ready or in a terminal state. This is what allows the initialization
+	// timeout logic to work correctly and return early - otherwise, users would have to wait the full init timeout
+	// before receiving a status update. The following are true:
+	// 1. Initializers may close the channel (because an initializer's job is to initialize the SDK!)
+	// 2. Synchronizers may close the channel (because an initializer might not be configured, or have failed)
+	// To ensure the channel is closed only once, we use a sync.Once wrapping the close() call.
+	readyOnce sync.Once
+
+	// These broadcasters are mainly to satisfy the existing SDK contract with users to provide status updates for
+	// the data source, data store, and flag change events. These may be different in fdv2, but we attempt to implement
+	// them for now.
+	broadcasters *broadcasters
+
+	// We hold a reference to the dataStoreStatusProvider because it's required for the public interface of the
+	// SDK client.
+	dataStoreStatusProvider interfaces.DataStoreStatusProvider
+}
+
+func NewFDv2(cfgBuilder subsystems.ComponentConfigurer[subsystems.DataSystemConfiguration], clientContext *internal.ClientContextImpl) (*FDv2, error) {
+
+	store := newStore(clientContext.GetLogging().Loggers)
+
+	bcasters := &broadcasters{
+		dataSourceStatus: internal.NewBroadcaster[interfaces.DataSourceStatus](),
+		dataStoreStatus:  internal.NewBroadcaster[interfaces.DataStoreStatus](),
+		flagChangeEvent:  internal.NewBroadcaster[interfaces.FlagChangeEvent](),
+	}
+
+	dataStoreUpdateSink := datastore.NewDataStoreUpdateSinkImpl(bcasters.dataStoreStatus)
+	clientContextCopy := *clientContext
+	clientContextCopy.DataStoreUpdateSink = dataStoreUpdateSink
+	clientContextCopy.DataSourceUpdateSink = store
```

Comment:
Some funkyness here just like in `FDv1` struct due to the way `clientContext` is passed between components during the configuration stage.